core/lib/server/pub.ts

```diff
@@ -1398,6 +1480,113 @@ export async function getPubsWithRelatedValuesAndChildren<
 						)
 					)
 			)
+			.with("stage_ms", (db) =>
+				db
+					.selectFrom("stage_memberships")
+					.$if(Boolean(props.userId), (qb) =>
+						qb
+							.where("stage_memberships.userId", "=", props.userId!)
+							.select([
+								"role",
+								"stageId as membId",
+								sql<MembershipType>`'stage'::"MembershipType"`.as("type"),
+							])
+					)
+					.$castTo<{
+						role: MembershipCapabilitiesRole;
+						membId: string;
+						type: MembershipType;
+					}>()
+			)
+			.with("pub_ms", (db) =>
+				db
+					.selectFrom("pub_memberships")
+					.$if(Boolean(props.userId), (qb) =>
+						qb
+							.where("pub_memberships.userId", "=", props.userId!)
+							.select([
+								"role",
+								"pubId as membId",
+								sql<MembershipType>`'pub'::"MembershipType"`.as("type"),
+							])
+					)
+					.$castTo<{
+						role: MembershipCapabilitiesRole;
+						membId: string;
+						type: MembershipType;
+					}>()
+			)
+			.with("community_ms", (db) =>
+				db
+					.selectFrom("community_memberships")
+					.$if(Boolean(props.userId), (qb) =>
+						qb
+							.where("community_memberships.userId", "=", props.userId!)
+							.where("community_memberships.communityId", "=", props.communityId)
+							.select([
+								"role",
+								"communityId as membId",
+								sql<MembershipType>`'community'::"MembershipType"`.as("type"),
+							])
+					)
+					// Add a fake community admin role when selecting without a userId
+					.$if(!Boolean(props.userId), (qb) =>
+						qb.select((eb) => [
+							sql<MemberRole>`${MemberRole.admin}::"MemberRole"`.as("role"),
+							eb.val(props.communityId).as("membId"),
+							sql<MembershipType>`'community'::"MembershipType"`.as("type"),
+						])
+					)
+
+					.$castTo<{
+						role: MembershipCapabilitiesRole;
+						membId: string;
+						type: MembershipType;
+					}>()
+			)
+			.with("memberships", (cte) =>
+				cte
+					.selectFrom("community_ms")
+					.selectAll("community_ms")
+					.$if(Boolean(props.userId), (qb) =>
+						qb
+							.union((qb) => qb.selectFrom("pub_ms").selectAll("pub_ms"))
+							.union((qb) => qb.selectFrom("stage_ms").selectAll("stage_ms"))
+							// Add fake community admin role when user is a superadmin
+							.union((qb) =>
+								qb
+									.selectFrom("users")
+									.where("users.id", "=", props.userId!)
+									.where("users.isSuperAdmin", "=", true)
+									.select((eb) => [
+										sql<MemberRole>`${MemberRole.admin}::"MemberRole"`.as(
+											"role"
+										),
+										eb.val(props.communityId).as("membId"),
+										sql<MembershipType>`'community'::"MembershipType"`.as(
+											"type"
+										),
+									])
+							)
+					)
+			)
+			.with("capabilities", (cte) =>
```

Comment:
wow this is very involved, thank you for the PR comment walkthrough!

I think I need to do someting similar for stages, here's what I had so far... https://github.com/pubpub/platform/pull/856 I was getting kind of stuck though, so I'll be referring to this! but also if you have any tips having just gone through this, I'd love them 😅

Reply:
it looks like you're pretty close and a similar but simpler approach should work for you! if you take the queries you've written and combine them into one or three CTEs in the `getStages` query, you should be able to `.innerJoin` your last cte (the deduplicated list of all authorized stageIds) to the main query and that should filter it. since you're writing my tests for me, i'd be happy to try that for you if you want!

Reply:
ooh okay lemme try that out first and I'll ping you if I run into problems! thanks for the guidance!

Reply:
whew your instructions worked perfectly! much simpler than what you had to do here 😅